src/sync/sync_user.cpp

```diff
@@ -0,0 +1,166 @@
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "sync_user.hpp"
+
+#include "impl/sync_metadata.hpp"
+#include "sync_manager.hpp"
+#include "sync_session.hpp"
+
+namespace realm {
+
+SyncUser::SyncUser(std::string refresh_token, std::string identity, bool is_admin)
+: m_state(State::Active)
+, m_is_admin(is_admin)
+, m_refresh_token(std::move(refresh_token))
+, m_identity(std::move(identity))
+{ }
+
+std::vector<std::shared_ptr<SyncSession>> SyncUser::all_sessions()
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    std::vector<std::string> keys_to_remove;
+    std::vector<std::shared_ptr<SyncSession>> sessions;
+    if (m_state == State::Error) {
+        return sessions;
+    }
+    for (auto it : m_sessions) {
+        if (auto ptr = it.second.lock()) {
+            if (ptr->is_valid()) {
+                sessions.emplace_back(std::move(ptr));
+                continue;
+            }
+        }
+        keys_to_remove.emplace_back(it.first);
+    }
+    for (auto key : keys_to_remove) {
+        // Clear any sessions that have been destroyed or fatally errored out.
+        m_sessions.erase(key);
+    }
+    return sessions;
+}
+
+std::shared_ptr<SyncSession> SyncUser::session_for_url(const std::string& url)
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    if (m_state == State::Error) {
+        return nullptr;
+    }
+    auto it = m_sessions.find(url);
+    if (it == m_sessions.end()) {
+        return nullptr;
+    }
+    auto locked = it->second.lock();
+    if (!locked) {
+        // Remove the session from the map, because it has fatally errored out or the entry is invalid.
+        m_sessions.erase(it);
+    }
+    return locked;
+}
+
+void SyncUser::update_refresh_token(std::string token)
+{
+    std::unique_lock<std::mutex> lock(m_mutex);
+    switch (m_state) {
+        case State::Error:
+            return;
+        case State::Active:
+            m_refresh_token = std::move(token);
+            break;
+        case State::LoggedOut: {
+            m_refresh_token = std::move(token);
+            m_state = State::Active;
+            // (Re)activate all pending sessions.
+            for (auto pair : m_waiting_sessions) {
+                if (auto ptr = pair.second.lock()) {
+                    m_sessions[pair.first] = ptr;
+                    lock.unlock();
```

Comment:
There was an issue with this snippet of code that I noticed at your prompting. I've rewritten it to be correct.